src/main/java/net/snowflake/client/jdbc/SnowflakeChunkDownloader.java

```diff
@@ -306,9 +307,18 @@ private void startNextDownloaders() throws SnowflakeSQLException {
 
       // try to reserve the needed memory
       long curMem = currentMemoryUsage.addAndGet(neededChunkMemory);
+      logger.debug(
+          "memoryLimit: {}, curMem: {}, retry: {}, maxRetry: {}",
+          memoryLimit,
+          curMem,
+          retry,
+          MAX_NUM_OF_RETRY);
       // no memory allocate when memory is not enough for prefetch
-      if (curMem > memoryLimit && nextChunkToDownload - nextChunkToConsume > 0) {
+      if (curMem > memoryLimit
+          && (nextChunkToDownload - nextChunkToConsume > 0
+              || (nextChunkToDownload == 0 && nextChunkToConsume == 0))) {
         // cancel the reserved memory and this downloader too
+        logger.debug("Not enough memory available for prefetch. Cancel reserved memory");
```

Comment:
good to add the values of `nextChunkToDownload` and `nextChunkToConsume` in the log.